### DISCUSSION TITLE
Class Data Sharing for reducing startup time and memory footprint

### DISCUSSION THREAD

**Discussion by pedroigor** [keycloak/keycloak#12409] [Archive]
To help reduce startup time and memory footprint, I would like to propose introducing a CDS archive into the distribution based on the classes loaded when running a JVM for a "naked" distribution.

Although we are not really interested in sharing the archive across multiple JVM processes running on a host, it should help to load classes faster. I'm still evaluating how much we should save.

There are several ways we can generate the CDS archive from Maven by using Quarkus CDS support.

We can also start doing this only for container images and not necessarily include the archive in the distribution. That should improve the experience when running as a container given the resource constraints. In this case, we would be creating the CDS archive into the base image and setting the necessary JVM options to enable CDS.

**Comment by bs-matil** [comment]
Sound very interesting, The feature page states this is experimental do you know if this is safe for production, or will this be an optional flag? 
https://quarkus.io/guides/maven-tooling#quarkus-package-pkg-package-config_quarkus.package.create-appcds

**Reply by pedroigor** [reply]
CDS is enabled by default and a production feature in the latest JVMs.

We would need to bump the JDK version in our side (e.g.: image) too and find the version we should be using because there were improvements over time around CDS.

**Comment by thomasdarimont** [comment]
I did some experiments recently and was not too impressed with the results. With ADS/CDS I could only shave off a few 100ms startup time.
See: https://gist.github.com/thomasdarimont/48e4b9e28be3ac02cf6cba33c62f3842

**Reply by pedroigor** [reply]
For me, it saves about 1s. Also, not this is about memory footprint too, something I did not check yet.

One key difference here is that I'm using JDK 13+. Can you try it using the same version?

Could you also try running locally (not as a container)?

**Reply by thomasdarimont** [reply]
I only tested this with JDK 17. Between Java 11 and Java 17 I could observe a decent reduction in startup time (~1s - 1.5s) from switching to JDK 17 alone.

Do you think it make sense to test the intermediate JDK versions as well? Most customers that I known only use a LTS version, e.g. JDK11 or JDK17 for running Keycloak in production.

But I can for sure try this with older JDKs as well.

**Reply by pedroigor** [reply]
I think we should track the latest releases to make sure the improvements are there and justify the investment.

In the worst case, we can create/update a guide with some instructions about how to optimize runtime using this approach?
